### PR TITLE
Improve as_json / to_json handling

### PR DESCRIPTION
**Pull request by ryabrody**
- change `to_json` on `Collection` to return just the items and not including the parent hash
- change `to_json` on `Collection` to support options
- change `_raw` on `Data` to return only hashes with symbols as keys

### CODE REVIEW

**Comment by kot-begemot** [comment]
:+1: looks good!

File: lib/lhs/concerns/data/json.rb [hunk 1]
```diff
@@ -5,8 +5,14 @@ class LHS::Data
   module Json
     extend ActiveSupport::Concern
 
-    def as_json(_options = {})
-      _data._raw.as_json
+    def as_json(options = {})
+      if _data._raw.is_a?(Array)
+        _data._raw.as_json(options)
+      elsif _data._raw[:items].present?
```
Comment (Averethel) [inline lib/lhs/concerns/data/json.rb]: That breaks as soon as we have a non collection object with key `items`. Can we check if it's a collection in a different way?

**Comment by Averethel** [comment]
I have two concerns regarding `to_json` on Collection:
1) `as_json` might break if we have an object with key `items`, which does not necessary need to be collection.
2) why don't we want to have a root key on the output JSON? I honestly prefer that to having either object or Array

Options support and symbolizing keys looks good.

File: lib/lhs/concerns/data/json.rb (hunk 1 above)
Comment (ryabrody) [inline lib/lhs/concerns/data/json.rb]: @Averethel thx for your feedback. 
Can you pls make an example for "non collection object with key items" ?

**Comment by ryabrody** [comment]
@Averethel thx for your feedback. Would the second point you came up with not be an "architectal change" which may would be too big for the change i wanted to achieve in this PR? I am also not sure why we remove the root key i think this would be a question for @pape-io. I personally only use it in specs examples like in the description of this PR: https://github.com/local-ch/lhs/pull/89.

**Comment by ryabrody** [comment]
@Averethel I covered now  your 1) point in a spec and check that the call of to_json on a non-collection Object with key `items` does not break.

**Comment by Averethel** [comment]
@ryabrody okay that looks better now.

I'd still like some input from @pape-io or @acoiro  as they have more knowledge on this library

File: lib/lhs/data.rb [hunk 2]
```diff
@@ -59,7 +59,7 @@ def respond_to_missing?(name, include_all = false)
   private
 
   def collection_proxy?(input)
-    !! (input.is_a?(Hash) && input[items_key]) || input.is_a?(Array) || _raw.is_a?(Array)
+    !! (input.is_a?(Hash) && self.class.present? && input[items_key]) || input.is_a?(Array) || _raw.is_a?(Array)
```
Comment (acoiro) [inline lib/lhs/data.rb]: Is it possible that class isn't defined?

File: lib/lhs/concerns/data/json.rb (hunk 1 above)
Comment (acoiro) [inline lib/lhs/concerns/data/json.rb]: I would wait for @pape-io, I don't know LHS well enough to say that we want this behaviour.

File: lib/lhs/data.rb [hunk 3]
```diff
@@ -105,6 +105,8 @@ def raw_from_json_string(input)
     json = JSON.parse(input)
     if json.is_a?(Hash)
       json.deep_symbolize_keys
+    elsif json.is_a?(Array)
+      json.map { |j| j.is_a?(Hash) ? j.deep_symbolize_keys : j }
```
Comment (acoiro) [inline lib/lhs/data.rb]: What about an array of an array of hashes? Or a hash containing arrays of hashes? We should probably do it recursively.

Comment (acoiro) [inline lib/lhs/data.rb]: Sorry `deep_symbolize_keys` is already doing the recursion for us ;-)

File: lib/lhs/concerns/data/json.rb (hunk 1 above)
Comment (pape-io) [inline lib/lhs/concerns/data/json.rb]: The name for the key for `items` and `limit` is defined here: https://github.com/local-ch/lhs/blob/master/lib/lhs/data.rb#L8

File: lib/lhs/concerns/data/json.rb [hunk 4]
```diff
@@ -5,8 +5,14 @@ class LHS::Data
...
+        _data._raw[:items].as_json(options)
```
Comment (pape-io) [inline lib/lhs/concerns/data/json.rb]: The name for the key for items and limit is defined here: https://github.com/local-ch/lhs/blob/master/lib/lhs/data.rb#L8

**Comment by pape-io** [comment]
Using options for `as_json` is actually a benefit

File: lib/lhs/data.rb (hunk 2 above)
Comment (pape-io) [inline lib/lhs/data.rb]: Why is this change required?

File: lib/lhs/concerns/data/json.rb (hunk 4 above)
Comment (pape-io) [inline lib/lhs/concerns/data/json.rb]: Otherwise this change would be an improvement.

File: spec/data/to_json_spec.rb [hunk 5]
```diff
@@ -27,11 +27,48 @@ class Record < LHS::Record
...
+      LHS::Data.new([{ foo: 'foo', bar: 'bar' }])
+    end
+
+    it 'converts with options to json' do
+      expect(collection.as_json(only: :foo)).to eq [{ 'foo' => 'foo' }]
+      expect(collection.to_json(only: :foo)).to eq "[{\"foo\":\"foo\"}]"
+    end
+  end
+
+  context 'non collection object with key items' do
```
Comment (pape-io) [inline spec/data/to_json_spec.rb]: Don't understand the context at the actually test. Lets check the test together.
What I don't understand especially is why `only: :foo` returns `666`, but the key of the collection is actually `items`

**Comment by pape-io** [comment]
I pointed out some stuff, some of it we have to check together. The rest looks good :+1:

**Comment by pape-io** [comment]
@Averethel, @ryabrody what do you mean with root key?
Something like:
```
feedback.as_json
{ feedback: {some_more_data: {}} }
```
?

File: lib/lhs/data.rb (hunk 2 above)
Comment (ryabrody) [inline lib/lhs/data.rb]: @pape-io 
When the input is not a Hash and either an Array. When it as example an Integer like in this spec example:
`rspec ./spec/data/to_json_spec.rb:69`
Then it returns this Error Message:
```
Module::DelegationError: LHS::Data#items_key delegated to self.class.items_key, but self.class is nil: #<LHS::Data:0x007f99b4e24690 @_raw={:items=>666}, @_parent=nil, @_record=nil>
```

File: lib/lhs/data.rb (hunk 3 above)
Comment (pape-io) [inline lib/lhs/data.rb]: Abbreviations violate principle of visibility.

File: lib/lhs/concerns/data/json.rb (hunk 4 above)
Comment (pape-io) [inline lib/lhs/concerns/data/json.rb]: @ryabrody As mentioned, the problem with this that all the other meta-data from the backend will not be forwarded or reachable anymore. Example:

```
Backend Response:
{
items: [1,2,3,4,5,6,7,8,9,10]
total: 100,
limit: 10,
offset: 0
}
LHS:
```
suggestions.as_json
#[1,2,3,4,5,6,7,8,9,10]
```
